etc/ffgen.c

```diff
@@ -37,97 +46,149 @@ void make_ff()
                 ch[j] = i;
                 num_ff++;
             }
+            if (d > 0 && d - 1 > max_deg)
+                max_deg = d - 1;
         }
     }
 }
 
-void emit_code(int header)
+unsigned int needed_bits(unsigned int x)
+{
+    if (x == 0)
+        return 0;
+    return 1 + needed_bits(x >> 1);
+}
+
+void emit_code(FILE * dest, int header)
 {
     unsigned i, j;
     if (header) {
-        printf("#ifndef GAP_FFDATA_H\n");
-        printf("#define GAP_FFDATA_H\n");
-        printf("\n");
-        printf("enum {\n");
-        printf("    NUM_SHORT_FINITE_FIELDS = %d\n", num_ff);
-        printf("};\n");
-        printf("\n");
-        printf("extern unsigned long SizeFF[NUM_SHORT_FINITE_FIELDS+1];\n");
-        printf("extern unsigned char DegrFF[NUM_SHORT_FINITE_FIELDS+1];\n");
-        printf("extern unsigned long CharFF[NUM_SHORT_FINITE_FIELDS+1];\n");
-        printf("\n");
-        printf("#endif // GAP_FFDATA_H\n");
+        fprintf(dest, "/* This file is generated by etc/ffgen.c during build"
+                      " do not edit */\n");
```

Comment:
So the changes in this file are difficult to review, because of those `printf(` -> `fprintf(dest, ` changes. I wonder: are those really necessary? What's the problem with the old approach, where `ffgen` simply was called twice, and its output was piped into the relevant files? Is it perhaps too slow now?